core/src/sparsity.jl

```diff
@@ -11,124 +11,34 @@ from the naming convention of this sparsity structure in the
-        has_index_out, idx_out = id_index(basin.node_id, id_out)
-
-        if has_index_in
-            jac_prototype[idx_in, idx_in] = 1.0
-        end
-
-        if has_index_out
-            jac_prototype[idx_out, idx_out] = 1.0
-        end
-
-        if has_index_in && has_index_out
-            jac_prototype[idx_in, idx_out] = 1.0
-            jac_prototype[idx_out, idx_in] = 1.0
-        end
-    end
-    return nothing
-end
-
-"""
-Method for nodes that do not contribute to the Jacobian
-"""
-function update_jac_prototype!(
-    jac_prototype::SparseMatrixCSC{Float64, Int64},
-    p::Parameters,
-    node::AbstractParameterNode,
-)::Nothing
-    node_type = nameof(typeof(node))
-
-    if !isa(
-        node,
-        Union{
-            Basin,
-            DiscreteControl,
-            FlowBoundary,
-            FractionalFlow,
-            LevelBoundary,
-            Terminal,
-        },
-    )
-        error(
-            "It is not specified how nodes of type $node_type contribute to the Jacobian prototype.",
-        )
-    end
-    return nothing
-end
-
-"""
-If both the unique node upstream and the nodes down stream (or one node further
-if a fractional flow is in between) are basins, then the downstream basin depends
-on the upstream basin(s) and affect the Jacobian as many times as there are downstream basins
-Upstream basins always depend on themselves.
+Add nonzeros for basins connected to eachother via 1 node and possibly a fractional flow node
 """
 function update_jac_prototype!(
     jac_prototype::SparseMatrixCSC{Float64, Int64},
-    p::Parameters,
-    node::Union{Pump, Outlet, TabulatedRatingCurve, UserDemand},
+    basin::Basin,
+    graph::MetaGraph,
 )::Nothing
-    (; basin, fractional_flow, graph) = p
-
-    for (i, id) in enumerate(node.node_id)
-        id_in = inflow_id(graph, id)
-
-        if hasfield(typeof(node), :is_pid_controlled) && node.is_pid_controlled[i]
-            continue
-        end
-
-        # For inneighbors only directly connected basins give a contribution
-        has_index_in, idx_in = id_index(basin.node_id, id_in)
-
-        # For outneighbors there can be directly connected basins
-        # or basins connected via a fractional flow
-        # (but not both at the same time!)
-        if has_index_in
-            jac_prototype[idx_in, idx_in] = 1.0
-
-            _, basin_idxs_out, has_fractional_flow_outneighbors =
-                get_fractional_flow_connected_basins(id, basin, fractional_flow, graph)
-
-            if !has_fractional_flow_outneighbors
-                id_out = outflow_id(graph, id)
-                has_index_out, idx_out = id_index(basin.node_id, id_out)
-
-                if has_index_out
-                    jac_prototype[idx_in, idx_out] = 1.0
+    for (idx_1, id) in enumerate(basin.node_id)
+        for id_neighbor in inoutflow_ids(graph, id)
+            for id_neighbor_neighbor in inoutflow_ids(graph, id_neighbor)
```

Comment:
At first I thought this may take a bit long since it is neighbors of neigbors, but with the way our graphs are structured it should be fine, and is quite simple to read.

Otherwise I was thinking of neighborhood functions but I don't think that's worth it.
https://juliagraphs.org/Graphs.jl/stable/algorithms/connectivity/#Graphs.neighborhood_dists-Union{Tuple{T},%20Tuple{U},%20Tuple{AbstractGraph{T},%20Integer,%20Any},%20Tuple{AbstractGraph{T},%20Integer,%20Any,%20AbstractMatrix{U}}}%20where%20{U%3C:Real,%20T%3C:Integer}